### PR TITLE
Rework tile-consumer-fuse-producers

### PR DESCRIPTION
**Pull request by chelini**
Initially, we used `tileConsumerAndFuseProducerGreedilyUsingSCFForOp` to tile and fuse element-wise operation with matmul-like ops. However, the use of `tileConsumerAndFuseProducerGreedilyUsingSCFForOp` is inconvenient as we cannot control it. Commit:
https://github.com/llvm/llvm-project/commit/9db7d4edd82ed5bb3867bbe6093ed2c63b18425a in LLVM exposes internal building blocks; thus, we can now implement our version of `tileConsumerAndFuseProducerGreedilyUsingSCFForOp` but retaining complete control on what gets tiled or fused.

We also provide a default, where the two-outermost parallel loops are tiled by 32 and fused. The PR works for simple element-wise op + matmul producers. They approach we use is similar in spirit to what IREE does (thanks Mahesh).

### CODE REVIEW

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 1]
```diff
@@ -25,108 +27,92 @@ using namespace mlir;
...
-static bool isStaticRange(Range loopRange) {
-  Optional<int64_t> offsetAsInt = getConstantIntValue(loopRange.offset);
-  if (!offsetAsInt)
-    return false;
-  Optional<int64_t> sizeAsInt = getConstantIntValue(loopRange.size);
-  if (!sizeAsInt)
-    return false;
-  Optional<int64_t> strideAsInt = getConstantIntValue(loopRange.stride);
-  return static_cast<bool>(strideAsInt);
+// TODO: improve checks here.
```
Comment (rengolin) [inline lib/TPP/TileConsumerAndFuseProducers.cpp:30]: Be more specific on the kind of checks that are needed, so someone else can do that later, too.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 2]
```diff
@@ -25,108 +27,92 @@ using namespace mlir;
...
-  FuseGenericOp(MLIRContext *context, ArrayRef<int64_t> tileSizes,
-                PatternBenefit benefit = 1)
-      : OpRewritePattern<linalg::GenericOp>(context, benefit),
-        tileSizes(tileSizes) {}
+static FailureOr<scf::SCFTilingResult>
+tileConsumer(RewriterBase &rewriter, TilingInterface consumer,
+             ArrayRef<int64_t> tileSizes) {
+  // If tileSizes is provided use them otherwise defaul to 1, 1, meaning fuse
+  // entire the two outermost loops.
+  SmallVector<int64_t> actualTileSizes = tileSizes.empty()
```
Comment (rengolin) [inline lib/TPP/TileConsumerAndFuseProducers.cpp:57]: I wonder if this should have a default value at all, or if every call should make sure to pass the tiling sizes, to make sure we don't have surprise results, for instance, if the user does pass `tileSizes` but for some buggy reason, it's empty.

I'd favour to assert if there are no sizes.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 3]
```diff
@@ -25,108 +27,92 @@ using namespace mlir;
...
+                                             ? SmallVector<int64_t>{1, 1}
+                                             : llvm::to_vector(tileSizes);
+  auto options = scf::SCFTilingOptions().setTileSizes(actualTileSizes);
+  FailureOr<scf::SCFTilingResult> tilingResult =
```
Comment (rengolin) [inline lib/TPP/TileConsumerAndFuseProducers.cpp:61]: Don't we need to check if the number of tiles sizes is compatible with the tiling we're trying to do? Say, if you pass more sizes than there are dimensions?

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 4]
```diff
@@ -25,108 +27,92 @@ using namespace mlir;
...
+  scf::SCFTileAndFuseResult tileAndFuseResult;
+  FailureOr<scf::SCFTilingResult> tilingResult =
+      tileConsumer(rewriter, consumer, tileSizes);
+  if (failed(tilingResult))
+    return rewriter.notifyMatchFailure(consumer,
+                                       "failed to tile base operation");
+  for (auto *tiledOp : tilingResult->tiledOps)
+    tileAndFuseResult.tiledAndFusedOps.insert(tiledOp);
+  tileAndFuseResult.loops = std::move(tilingResult->loops);
+  for (const auto &result : llvm::enumerate(
```
Comment (rengolin) [inline lib/TPP/TileConsumerAndFuseProducers.cpp:81]: Use auto tuple notation to avoid `get<N>` and `zip_equal`.

File: test/Passes/tile-and-fuse.mlir [hunk 5]
```diff
@@ -0,0 +1,100 @@
+// RUN: tpp-opt %s -split-input-file -tile-consumer-and-fuse-producers -cse | FileCheck %s
+
+func.func @matmul_sequence_fusion(%arg0: tensor<32x64xf32>, %arg1: tensor<64x32xf32>,
```
Comment (adam-smnk) [inline test/Passes/tile-and-fuse.mlir:3]: I guess we don't fuse these yet - maybe add a simple check or a comment so that the test isn't completely empty

File: test/Passes/tile-and-fuse.mlir [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    %arg2: tensor<32x32xf32>, %arg3: tensor<32x64xf32>, %arg4: tensor<32x64xf32>,
+    %arg5: tensor<64x32xf32>, %arg6: tensor<32x32xf32>) -> tensor<32x32xf32> {
+  %0 = linalg.matmul ins(%arg0, %arg1 : tensor<32x64xf32>, tensor<64x32xf32>)
+    outs(%arg2 : tensor<32x32xf32>) -> tensor<32x32xf32> // [M, N0] * [N0, N1]
+  %1 = linalg.matmul ins(%0, %arg3 : tensor<32x32xf32>, tensor<32x64xf32>)
+    outs(%arg4 : tensor<32x64xf32>) -> tensor<32x64xf32> // [M, N1] * [N1, N2]
+  %2 = linalg.matmul ins(%1, %arg5 : tensor<32x64xf32>, tensor<64x32xf32>)
+    outs(%arg6 : tensor<32x32xf32>) -> tensor<32x32xf32> // [M, N2] * [N2, N3]
+  return %2 : tensor<32x32xf32>
+}
```
Comment (rengolin) [inline test/Passes/tile-and-fuse.mlir:14]: No check lines for this one?

File: test/Passes/tile-and-fuse.mlir [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+// CHECK: #[[MAP2:.+]] = affine_map<(d0, d1, d2, d3) -> (d1, d2)>
+// CHECK: #[[MAP3:.+]] = affine_map<(d0, d1) -> (d0, d1)>
+// CHECK: func.func @matmul_eletwise(
+// CHECK-SAME:  %[[ARG0:.+]]: tensor<4x4x32x32xf32>, %[[ARG1:.+]]: tensor<4x4x32x32xf32>,
+// CHECK-SAME:  %[[ARG2:.+]]: tensor<4x4x32x32xf32>)
+// CHECK: %[[C4:.+]] = arith.constant 4 : index
+// CHECK: %[[C0:.+]] = arith.constant 0 : index
+// CHECK: %[[C1:.+]] = arith.constant 1 : index
+// CHECK: %[[CST:.+]] = arith.constant 0.000000e+00 : f32
+// CHECK: %[[LOOP:.+]] = scf.for %[[ARG3:.+]] = %[[C0]] to %[[C4]] step %[[C1]] iter_args(%[[ARG4:.+]] = %[[ARG2]]) -> (tensor<4x4x32x32xf32>) {
```
[the remark this answers is not in the document]
Comment (chelini) [reply test/Passes/tile-and-fuse.mlir:95]: scf.parallel cannot be use at tensor level. Yesterday, we discussed this in the chat. The option is to look at other parallel construct to have parallelism at tensor level like `scf.foreach_thread` or raise it back to parallel.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 2 above)
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:57]: I am not sure either. I was thinking this in the context of having a default pipeline.

Comment (adam-smnk) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:57]: In the default pipeline we should try to adjust the sizes dynamically. But in the context of this pass on its own, it might be good QoL to have sth that works out of the box.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 4 above)
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:81]: C++ 17 does not support nesting in structured bindings. But I can replace llvm::zip with llvm::zip_equal

File: test/Passes/tile-and-fuse.mlir (hunk 7 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply test/Passes/tile-and-fuse.mlir:95]: ah, right, sorry.

Where will the tile `{1, 1}` make parallel loops, then?

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 2 above)
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:58]: The way it's done, we can introduce bugs by passing an empty container. So, if we want a default value, we need to make it default on the function declaration (`arg = {1,1}`), if not, we assert if it's empty.

I think we should start asserting, and only add a default when we're certain of the default we actually want. It's not clear to me yet that this is the default.

File: test/Passes/tile-and-fuse.mlir (hunk 7 above)
[the remark this answers is not in the document]
Comment (chelini) [reply test/Passes/tile-and-fuse.mlir:95]: We cannot make parallel loops at tensor level. {1, 1} is a simple default that will tile with 1 the two outermost loops.

[the remark this answers is not in the document]
Comment (chelini) [reply test/Passes/tile-and-fuse.mlir:95]: If is confusing or surprising we can just pass tile sizes to this pass.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 2 above)
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:58]: Ok, so we prefer passing tile sizes to the pass and avoid default for now, right? {1, 1} works relatively well for matmul + elemenwise because the output of the matmul is 2d, so you end up with a reduction + scalar and 2 parallel loops. Or if the matmul is packed with a 2d element wise and a brgemm op.

**Comment by chelini** [comment]
Ok let me think more about this. Thanks for the comments.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 2 above)
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:58]: Not tile sizes to the pass, to the function that calls this one. Ie. just move the default out of this "generic" function, so that its usage is clear and safe.

File: test/Passes/tile-and-fuse.mlir (hunk 7 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply test/Passes/tile-and-fuse.mlir:95]: No tile sizes to the pass... We need to discover this on our own, but not in a generic helper function.

The question here was: there's a claim that tiling `{1,1}` leads to `scf.parallel`, but the tests don't show this. As you said, what you have is a pair of `scf.for` that cannot be _parallel_ because they're on tensors, but it's a pattern that, later, can be converted to `scf.parallel` on memrefs.

I think it's clear that this should not be done in this PR, but probably a new issue to add this as a later pass on another PR.

File: test/Passes/tile-and-fuse-default.mlir [hunk 8]
```diff
@@ -0,0 +1,238 @@
+// RUN: tpp-opt %s -split-input-file -tile-consumer-and-fuse-producers -cse | FileCheck %s
+
+// CHECK-NOT: scf.for
```
Comment (rengolin) [inline test/Passes/tile-and-fuse-default.mlir:3]: Nit: better to match and put the CHECK line after the function like the others, to make it easier to know which op it applies to.

Also, add a check for the `func.func`, to make sure we're in the right place. Since this is a split-file, we may add tests before and the check will always pass.

File: test/Passes/tile-and-fuse-default.mlir [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+// CHECK-DAG: #[[MAP:.+]] = affine_map<() -> ()>
+// CHECK: func.func @matmul_sequence_fusion
+// CHECK-DAG: %[[C32:.+]] = arith.constant 32 : index
+// CHECK-DAG: %[[C0:.+]] = arith.constant 0 : index
+// CHECK-DAG: %[[C1:.+]] = arith.constant 1 : index
+// CHECK: %[[LOOP:.+]] = scf.for %{{.+}} = %[[C0]] to %[[C32]] step %[[C1]]
+// CHECK-NEXT: %[[LOOP1:.+]] = scf.for %{{.+}} = %[[C0]] to %[[C32]] step %[[C1]]
+// CHECK-COUNT-3: linalg.matmul
```
Comment (rengolin) [inline test/Passes/tile-and-fuse-default.mlir:127]: What are the types for these linalg ops?

Comment (chelini) [reply test/Passes/tile-and-fuse-default.mlir:129]: the generic are scalar here. Basically we have:
```
for ( M
  for (N
    linalg.matmul (only on the reduction loop)
    linalg.generic (scalar op)
```

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 10]
```diff
@@ -140,10 +272,30 @@ struct TileConsumerAndFuseProducers
     linalg::registerTilingInterfaceExternalModels(registry);
   }
   void runOnOperation() override {
+    func::FuncOp func = getOperation();
+    transform::TrivialPatternRewriter rewriter(&getContext());
+    llvm::SmallDenseSet<Operation *> fusedOps;
+    func->walk([&](linalg::LinalgOp linalgOp) {
+      if (linalg::isElementwise(linalgOp) && hasMatmulLikeProducer(linalgOp)) {
+        SmallVector<int64_t> actualTileSizes = llvm::to_vector(this->tileSizes);
+        // Default tiling scheme, always legal to tile by 1.
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:277]: I believe in this case it is a good default. If you have a matmul follow by an element-wise I think it is good to fuse and tile along the parallel dimensions. `tileSizes` comes from the pass option, only if no options are passed we use the default {1, 1}, otherwise we use the tile sizes provided by the user.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 11]
```diff
@@ -14,119 +14,251 @@
...
+  std::optional<OpOperand *> destinationIterArg;
+  auto loopIt = loops.rbegin();
+  while (auto iterArg = source->get().dyn_cast<BlockArgument>()) {
+    scf::ForOp loop = *loopIt;
+    if (iterArg.getOwner()->getParentOp() != loop)
+      break;
+    source = &loop.getOpOperandForRegionIterArg(iterArg);
+    loopIt++;
+  }
+  if (loopIt == loops.rend())
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: Yes, left over.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 12]
```diff
@@ -14,119 +14,251 @@
...
+      Operation *producer = operand.get().getDefiningOp();
+      if (!producer || !isa<linalg::LinalgOp>(producer) ||
+          fusableProducers.count(producer) ||
+          !canBeTiledWithCurrentSpec(producer, tileSizes) ||
+          alreadyFusedOps.count(producer) || producer->getNumResults() != 1 ||
+          !hasAllUsersInWorklist(producer, fusableProducers))
+        continue;
+      LLVM_DEBUG(llvm::dbgs()
+                 << "WORKLIST INSERT PRODUCER: " << producer << "\n");
+      worklist.push_back(producer);
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:161]: Yes, as long as all the ops have users in the current worklist.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 13]
```diff
@@ -14,119 +14,251 @@
...
+       tileIdx++) {
+    // Allow tiling only on parallel loop iterators.
+    if (!linalg::isParallelIterator(loopIteratorTypes[tileIdx]))
+      return false;
+    std::optional<int64_t> sizeRangeAtIdx =
+        getSizeRangeAtIdx(iterationDomain, tileIdx);
+    // Non constant range. Bail out and do not add the op as candidate.
+    if (!sizeRangeAtIdx)
+      return false;
+    // Corner case tile equals 0.
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:95]: User can pass tileSizes like {0, 0} which are wrong. `Range` is the loop range of the scf.for generated by tiling. By construction of scf.for the range is positive.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 3 above)
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: If the tile sizes are not correct we should not get there. I will put an assert.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 14]
```diff
@@ -14,119 +14,251 @@
...
 }
 
-struct FuseGenericOp : public OpRewritePattern<linalg::GenericOp> {
-  FuseGenericOp(MLIRContext *context, ArrayRef<int64_t> tileSizes,
-                PatternBenefit benefit = 1)
-      : OpRewritePattern<linalg::GenericOp>(context, benefit),
-        tileSizes(tileSizes) {}
+// Return the range for the iteration domain at position `idx` return nullopt if
+// the range cannot be statically computed.
+static std::optional<int64_t> getSizeRangeAtIdx(ArrayRef<Range> iterationDomain,
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: yes, good point.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:277]: If we don't like a default I can remove it.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:161]: I can add a test case.

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:161]: Is this align with what you had in mind: 0f0a6fa11841152d3fb2abea3f125be4fe33ea68 ?

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:95]: I see, than this isn't a corner case, just input validation, which should be done at input level.

In this case, the "input" is a pass option, so this should be done in the pass itself. Is there a way to validate the pass options at the table-gen level? If not, than this must be done at the constructor level.

And if the sizes cannot be negative, than why not use `uint64_t` instead? Or even `uint32_t`, `uint16_t` etc as we're not expecting 4GB tile sizes.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:161]: Yes, thanks!

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:277]: No, we want a default, I just think that `{1, 1}` isn't a good default. Tiling a matmul for each element isn't a good idea for almost any hardware.

According to libxsmm experiments, the meaningful default tile sizes are 32 and 64. IIUC, less than this, we don't fill the cache lines, more than that we start clashing lines due to associativity.

In that sense, the default value should be the minimum we recommend doing (in libxsmm, so 32), and only tile lower than that if the pass option is set.

In your testing, you can pass the option `{1, 1}` to the larger test (all alternatives) and add a small test with the default (and check that it's 32).

File: test/Passes/tile-and-fuse-default.mlir [hunk 15]
```diff
@@ -0,0 +1,274 @@
+// RUN: tpp-opt %s -split-input-file -tile-consumer-and-fuse-producers -cse | FileCheck %s
```
Comment (rengolin) [inline test/Passes/tile-and-fuse-default.mlir:1]: Add `tileSize={1,1}` here and create a new small test for the default of 32 elsewhere.

File: test/Passes/tile-and-fuse-default.mlir (hunk 9 above)
Comment (rengolin) [reply test/Passes/tile-and-fuse-default.mlir:129]: Can you just add the types to the CHECK lines, to make it clear what's going on?

I'm guessing `CHECK-COUNT` can cope with regexes, but if not, split into one with regex and the rest as a COUNT.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: The tiling works with int64_t so I would prefer to be consistent. Why do you want to validate in the constructor? What exactly do you expect the pass constructor to do if it sees a tile of 0?

[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: Re: `int64_t`, makes sense.

Re: Constructor, there should be a way to validate user input (ie. pass option) that doesn't involve barfing in the middle of the execution. We need to tell users they chose the wrong option and why.

If there's no decent way of doing that, then an early `return` in `runOnOperation` if the tiles are "wrong" is the first decent spot.

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: It seems that in linalg/scf a tile of '0' means skip this dimension and do not tile it. Maybe we can follow a similar approach,  what do you think?

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: In that case '0' is valid, just means "skip this dim"

[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: > In that case '0' is valid, just means "skip this dim"

That sounds like a good approach!

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp]: See: 48d629f1c3852dedb437a6d2ad13d1f7a660a2e3

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 16]
```diff
@@ -140,10 +272,30 @@ struct TileConsumerAndFuseProducers
     linalg::registerTilingInterfaceExternalModels(registry);
   }
   void runOnOperation() override {
+    func::FuncOp func = getOperation();
+    transform::TrivialPatternRewriter rewriter(&getContext());
+    llvm::SmallDenseSet<Operation *> fusedOps;
+    func->walk([&](linalg::LinalgOp linalgOp) {
+      if (linalg::isElementwise(linalgOp) && hasMatmulLikeProducer(linalgOp)) {
```
[the remark this answers is not in the document]
Comment (adam-smnk) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:279]: Nothing special but it is clearly capable of fusing chain of matmuls or other ops. So, it should be able to kick in without an element-wise op. Or will that break some current assumptions?

[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:279]: It seems `hasMatmulLikeProducer` only checks for immediate producers, so a `matmul -> ew1 -> ew2` won't fuse `ew2` since it doesn't have `matmul` producers (and this check will fail).

The test `@matmul_sequence_fusion` below shows that the final `generic (addf)` doesn't get fused with the bundle, because its producer is a `generic (maxf)`. But if the tiling spec is the same, then there's no reason for that not to be fused, even if the producer isn't a matmul.

[the remark this answers is not in the document]
Comment (adam-smnk) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:279]: It probably could iterate greedily over the whole chain of ops and only stop when you cannot fuse the next producer/consumer. As there could be some pitfalls to this approach, I guess it could be investigated separately and we can keep the simpler anchor approach for this PR.

[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:279]: I think this is complex enough that deserves to be in a separate PR, and with more discussion around it.

**Comment by chelini** [comment]
Found out that some tests are wrong:
```
#map = affine_map<() -> ()>
#map1 = affine_map<(d0, d1) -> (d0, d1)>
module {
  func.func @matmul_sequence_fusion(%arg0: tensor<32x64xf32>, %arg1: tensor<64x32xf32>, %arg2: tensor<32x32xf32>, %arg3: tensor<32x64xf32>, %arg4: tensor<32x64xf32>, %arg5: tensor<64x32xf32>, %arg6: tensor<32x32xf32>) -> tensor<32x32xf32> {
    %c32 = arith.constant 32 : index
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %cst = arith.constant 0.000000e+00 : f32
    %0 = scf.for %arg7 = %c0 to %c32 step %c1 iter_args(%arg8 = %arg6) -> (tensor<32x32xf32>) {
      %3 = scf.for %arg9 = %c0 to %c32 step %c1 iter_args(%arg10 = %arg8) -> (tensor<32x32xf32>) {
        %extracted_slice = tensor.extract_slice %arg0[%arg7, 0] [1, 64] [1, 1] : tensor<32x64xf32> to tensor<1x64xf32>
        %extracted_slice_0 = tensor.extract_slice %arg2[%arg7, 0] [1, 32] [1, 1] : tensor<32x32xf32> to tensor<1x32xf32>
        %4 = linalg.matmul ins(%extracted_slice, %arg1 : tensor<1x64xf32>, tensor<64x32xf32>) outs(%extracted_slice_0 : tensor<1x32xf32>) -> tensor<1x32xf32>
        %extracted_slice_1 = tensor.extract_slice %arg4[%arg7, 0] [1, 64] [1, 1] : tensor<32x64xf32> to tensor<1x64xf32>
        %5 = linalg.matmul ins(%4, %arg3 : tensor<1x32xf32>, tensor<32x64xf32>) outs(%extracted_slice_1 : tensor<1x64xf32>) -> tensor<1x64xf32>
        %extracted_slice_2 = tensor.extract_slice %arg5[0, %arg9] [64, 1] [1, 1] : tensor<64x32xf32> to tensor<64x1xf32>
        %extracted_slice_3 = tensor.extract_slice %arg10[%arg7, %arg9] [1, 1] [1, 1] : tensor<32x32xf32> to tensor<1x1xf32>
        %6 = linalg.matmul ins(%5, %extracted_slice_2 : tensor<1x64xf32>, tensor<64x1xf32>) outs(%extracted_slice_3 : tensor<1x1xf32>) -> tensor<1x1xf32>
        %extracted_slice_4 = tensor.extract_slice %6[0, 0] [1, 1] [1, 1] : tensor<1x1xf32> to tensor<f32>
        %7 = linalg.generic {indexing_maps = [#map], iterator_types = []} outs(%extracted_slice_4 : tensor<f32>) {
        ^bb0(%out: f32):
          %8 = arith.maxf %out, %cst : f32
          linalg.yield %8 : f32
        } -> tensor<f32>
        %inserted_slice = tensor.insert_slice %7 into %arg10[%arg7, %arg9] [1, 1] [1, 1] : tensor<f32> into tensor<32x32xf32>
        scf.yield %inserted_slice : tensor<32x32xf32>
      }
      scf.yield %3 : tensor<32x32xf32>
    }
    %1 = linalg.generic {indexing_maps = [#map1], iterator_types = ["parallel", "parallel"]} outs(%0 : tensor<32x32xf32>) {
    ^bb0(%out: f32):
      %3 = arith.maxf %out, %cst : f32
      linalg.yield %3 : f32
    } -> tensor<32x32xf32>
    %2 = linalg.generic {indexing_maps = [#map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%arg6 : tensor<32x32xf32>) outs(%1 : tensor<32x32xf32>) {
    ^bb0(%in: f32, %out: f32):
      %3 = arith.addf %in, %out : f32
      linalg.yield %3 : f32
    } -> tensor<32x32xf32>
    return %2 : tensor<32x32xf32>
  }
}```

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 17]
```diff
@@ -14,119 +14,421 @@
...
+getTileBitVectorConfig(ArrayRef<int64_t> tileSizes, size_t rootConsumerParallelLoops) {
+  assert(tileSizes.size() <= rootConsumerParallelLoops);
+  llvm::SmallBitVector tileConfig(rootConsumerParallelLoops, false);
+  for (size_t idx : llvm::seq<size_t>(0, tileSizes.size()))
+    if (tileSizes[idx] != 0)
+      tileConfig.set(idx);
+  return tileConfig;
+}
+
+static void _and(llvm::SmallBitVector &parallelLoops,
```
Comment (adam-smnk) [inline lib/TPP/TileConsumerAndFuseProducers.cpp:138]: I think you could just use the built-in `SmallBitVector &operator&=` instead. Or at least give this function a bit more expressive name.

**Review by adam-smnk — APPROVED** [review]
Looks fine to me.
In general the approach seems reasonable and not too greedy. Plus there's some potential to generalize it a bit further e.g., fusing chain of elementwise ops. But let's figure it out separately.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 18]
```diff
@@ -140,10 +272,30 @@ struct TileConsumerAndFuseProducers
...
   }
   void runOnOperation() override {
+    func::FuncOp func = getOperation();
+    transform::TrivialPatternRewriter rewriter(&getContext());
+    llvm::SmallDenseSet<Operation *> fusedOps;
+    func->walk([&](linalg::LinalgOp linalgOp) {
+      if (linalg::isElementwise(linalgOp) && hasMatmulLikeProducer(linalgOp)) {
+        SmallVector<int64_t> actualTileSizes = llvm::to_vector(this->tileSizes);
+        // Default tiling scheme tile by 32.
+        if (actualTileSizes.empty())
```
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:446]: You merged without addressing this issue. Please create a new issue to change this behaviour.

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:446]: You want to move the default in the constructor?

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:446]: the default of {32,32} is specific for this pattern.

[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:446]: No, the tile size of 32 is a "good enough" default for any pattern, because it's what libxsmm prefers to run. Anything smaller than this, by default, should NOT be tiled (unless forced via this option).
